File: Top 100 Codes/06-Advanced Codes related to Arrays/15-Find if there is any subarray with sum equal to 0/Findifthereisanysubarraywithsumequalto0.cpp
#include <iostream>
#include <unordered_set>
#include <vector>
using namespace std;

bool hasZeroSumSubarray(const vector<int>& nums) {
    unordered_set<int> sumSet; // To store cumulative sums
    int sum = 0;

    for (int i = 0; i < nums.size(); i++) {
        sum += nums[i];

        // If sum is 0 or already exists in the set, a zero-sum subarray exists
        if (sum == 0 || sumSet.find(sum) != sumSet.end()) {
            return true;
        }

        // Insert the current sum into the set
        sumSet.insert(sum);
    }

    return false; // No subarray with sum 0 found
}

int main() {
    vector<int> arr = {4, 2, -3, 1, 6};

    if (hasZeroSumSubarray(arr)) {
        cout << "Yes, there is a subarray with sum 0.\n";
    } else {
        cout << "No, there is no subarray with sum 0.\n";
    }

    return 0;
}

File: Top 100 Codes/06-Advanced Codes related to Arrays/15-Find if there is any subarray with sum equal to 0/Findifthereisanysubarraywithsumequalto0.py
def has_zero_sum_subarray(arr):
    sum_set = set()
    current_sum = 0

    for i in range(len(arr)):
        current_sum += arr[i]

        # If current sum is 0 or already exists in set
        if current_sum == 0 or current_sum in sum_set:
            return True

        sum_set.add(current_sum)

    return False

# Example usage
arr = [4, 2, -3, 1, 6]
if has_zero_sum_subarray(arr):
    print("Yes, there is a subarray with sum 0.")
else:
    print("No, there is no subarray with sum 0.")
